include/boost/json/impl/array.ipp

```diff
@@ -119,25 +240,26 @@ array(
     storage_ptr sp)
     : sp_(std::move(sp))
 {
-    undo_construct u(*this);
-    reserve(count);
-    while(impl_.size() < count)
+    if(count == 0)
     {
-        ::new(
-            impl_.data() +
-            impl_.size()) value(sp_);
-        impl_.size(impl_.size() + 1);
+        t_ = &empty_;
+        return;
+    }
+    t_ = table::allocate(
+        count, sp_);
+    t_->size = count;
+    auto p = data();
+    do
+    {
+        ::new(p++) value(sp_);
```

Comment:
Something that has applications outside of just `array` (you might not like this because it would be private interface): we could just `memcpy` the `storage_ptr` and at the end just have a single branch which checks `is_counted`, and if that returns `true` we increment the ref-count by `count`. It would kill a branch in each construction (thus optimizing the non-counted case) and we would only have to do one atomic add instead of `count` (atomic adds are expensive -- `lock add QWORD PTR[rax], 1` takes 3x as many cycles as `add QWORD PTR[rax], 1`), which would optimize the counted case.

Reply:
This would especially help in `value_stack`. Since we don't do deep copies, we can check `is_counted` and increment the ref-counter only when `push_array`, `push_object` and `release` are called.